ui: updates useSWR hooks to include clusterId in key

kyle-a-wong opened the pull request:
Previously, useSWR hooks were usings keys that didn't include cluster ids. This worked fine for DB console, but resulted in rendering bugs in cc-console. Since cc-console allows users to switch between clusters, users would temporarily see database and table metadata from previously views clusters until the useSWR hook refreshed the data.

Now that these hooks calls have been updated to include cluster id, switching between clusters will only show data relevant to that cluster.

Fixes: https://cockroachlabs.atlassian.net/browse/CC-30169
Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136261)
<!-- Reviewable:end -->

kyle-a-wong committed:
ui: create wrappers for useSWR to include cluster id

Previously, useSWR hooks were usings keys that didn't
include cluster ids. This worked fine for DB console, but
resulted in rendering bugs in cc-console. Since cc-console
allows users to switch between clusters, users would
temporarily see database and table metadata from previously
views clusters until the useSWR hook refreshed the data.

Now, these components use a new useSwrWithClusterId hook
that wraps useSWR and adds the cluster id to the key. This
hook should be preferred over the unwrapped hooks to
ensure that cc-console doesn't erroneously render the
wrong cluster's data.

Fixes: https://cockroachlabs.atlassian.net/browse/CC-30169
Epic: none
Release note: none

xinhaoz commented on pkg/ui/workspaces/cluster-ui/src/api/databases/tableMetaUpdateJobApi.ts:
```diff
@@ -71,7 +72,7 @@ export const convertTableMetaUpdateJobFromServer = (
 // 10 seconds otherwise.
 export const useTableMetaUpdateJob = () => {
   const { data, isLoading, mutate } = useSWR(
-    "tableMetaUpdateJob",
```
We can pass an array of keys to useSWR:
```
  const { clusterId } = useContext(ClusterDetailsContext);
  const { data, isLoading, mutate } = useSWR(
    [clusterId, "tableMetaUpdateJob"],
    getTableMetaUpdateJobInfo,
    {
....
```

xinhaoz commented on pkg/ui/workspaces/cluster-ui/src/api/nodesApi.ts:
```diff
@@ -26,10 +27,9 @@ export type NodeStatus = {
+  const { isTenant } = useContext(ClusterDetailsContext);
   const { data, isLoading, error } = useSWR(
-    NODES_PATH,
+    useSwrKeyWithClusterId({ name: "nodesUI" }),
```
I think we should try to avoid calling a hook directly as a function argument.

xinhaoz reviewed (commented):
I'm not sure about the usage of this new `useSwrKeyWithClusterId` hook. I think we should just use `useContext(ClusterDetailsContext)` directly and pass the clusterId to `useSwr` - you can pass an array of keys as an arg to useSwr. I don’t think it’s that much more code to do this and it’s more direct. Alternatively we could make `useSwrWithClusterId` replace the usages of useSWR by making it combine `useContext(ClusterDetailsContext)` + `useSwr` in the body. LMK what you think!

kyle-a-wong commented:
> I'm not sure about the usage of this new `useSwrKeyWithClusterId` hook. I think we should just use `useContext(ClusterDetailsContext)` directly and pass the clusterId to `useSwr` - you can pass an array of keys as an arg to useSwr. I don’t think it’s that much more code to do this and it’s more direct. Alternatively we could make `useSwrWithClusterId` replace the usages of useSWR by making it combine `useContext(ClusterDetailsContext)` + `useSwr` in the body. LMK what you think!

@xinhaoz I decided to go with the suggestion of making a new `useSwrWithClusterId` hook to wrap `useSWR`(and `useSWRImmutable`).  I think this is less error prone than require everyone to make sure they remember to include the clusterId in the key. If they really don't need the cluster id they can still use `useSWR`, but `useSwrWithClusterId` should be the hook most people reach for first.

xinhaoz reviewed (approved):
Nice, thanks for the changes!

kyle-a-wong commented on pkg/ui/workspaces/cluster-ui/src/api/databases/tableIndexesApi.ts:
```diff
@@ -49,8 +48,8 @@ export const useTableIndexStats = ({
-  const { data, isLoading, error, mutate } = useSWR(
-    ["tableIndexes", dbName, tableName],
+  const { data, isLoading, error, mutate } = useSwrWithClusterId(
+    { name: "tableIndexes", dbName, tableName },
```
I originally changed it when i had the old hook for adding the clusterId. Mainly I just thought it would be better to have a standardized way that we define our keys (which was using objects instead of arrays). After changing the hook i just didn't change this one back.

kyle-a-wong commented:
tftr!

bors r+
